Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,213 @@
+      map_smul' := fun s _ => show l _ = s • l _ by rw [← smul_tmul', map_smul] }
+  map_add' := fun _ _ => LinearMap.ext fun _ => show l _ = l _ + l _ by rw [tmul_add, map_add]
+  map_smul' := fun r _ => LinearMap.ext fun _ => show l _ = l (r • _ ⊗ₜ _) by
+    rw [← smul_tmul, smul_tmul']
+
+attribute [aesop unsafe] add_comm in
+/-- curry a bilinear map into a map from tensor product -/
+@[simps]
+def uncurry' {X' : ModuleCat.{v} R} {Y : ModuleCat.{v} Sᵐᵒᵖ} (l : X' →ₗ[R] (X →ₗ[Sᵐᵒᵖ] Y)) :
+    (X ⊗[R] X') →ₗ[Sᵐᵒᵖ] Y :=
+  let L : (X ⊗[R] X') →+ Y := (addConGen _).lift (FreeAddMonoid.lift fun p => l p.2 p.1) <|
+    AddCon.addConGen_le <| by
+      rintro _ _ (_|_|_|_|_|_) <;> refine' (AddCon.ker_rel _).2 _ <;> try aesop
+      simpa only [FreeAddMonoid.lift_eval_of, l.map_smul] using
+        Eq.symm <| LinearMap.bimodule_smul_apply (MulOpposite.op _) (l _) _
+  { L with
+    map_smul' := fun _ z => z.induction_on
+      (by aesop) (fun _ _ => LinearMap.map_smul _ _ _) (by aesop) }
+
+@[simp high]
+lemma uncurry'_apply_tmul {X' : ModuleCat.{v} R} {Y : ModuleCat.{v} Sᵐᵒᵖ}
+    (l : X' →ₗ[R] (X →ₗ[Sᵐᵒᵖ] Y))
+    (x : X) (x' : X') : uncurry' l (x ⊗ₜ x') = l x' x := rfl
+
+variable (R S X)
+/-- The tensor functor is left adjoint to the hom functor. -/
+noncomputable def tensorHomAdjunction : tensorFunctor R S X ⊣ homFunctor R S X :=
+    Adjunction.mkOfHomEquiv
+    { homEquiv := fun _ _ =>
+      { toFun := curry'
+        invFun := uncurry'
+        left_inv := fun _ => LinearMap.ext fun x => x.induction_on (by aesop) (by aesop) (by aesop)
+        right_inv := fun _ => LinearMap.ext fun _ => LinearMap.ext fun _ => rfl }
+      homEquiv_naturality_left_symm := fun {_ _} _ _ _ => by
+        refine' LinearMap.ext fun x => x.induction_on _ _ _ <;> aesop
+      homEquiv_naturality_right := by aesop }
+
+noncomputable instance : IsLeftAdjoint (tensorFunctor R S X) :=
+⟨_, tensorHomAdjunction _ _ _⟩
+
+noncomputable instance : IsRightAdjoint (homFunctor R S X) :=
+⟨_, tensorHomAdjunction _ _ _⟩
+
+noncomputable instance : Limits.PreservesColimits (tensorFunctor R S X) :=
+Adjunction.leftAdjointPreservesColimits <| tensorHomAdjunction _ _ _
+
+noncomputable instance : Limits.PreservesLimits (homFunctor R S X) :=
+Adjunction.rightAdjointPreservesLimits <| tensorHomAdjunction _ _ _
+
+instance : Functor.PreservesEpimorphisms (tensorFunctor R S X) :=
+inferInstance
+
+end ModuleCat
+
+namespace TensorProduct
+
+variable (R' : Type u) [CommRing R']
+variable {M N : Type v} [AddCommGroup M] [AddCommGroup N]
+variable [Module R' M] [Module R' N]
+
+instance int_mop_module : Module R'ᵐᵒᵖ N :=
+    Module.compHom N ((RingHom.id R').fromOpposite fun _ _ => mul_comm _ _)
+
+instance int_mop_bimod {N' : Type v} [AddCommGroup N'] [Module R' N'] :
+    SMulCommClass R' ℤᵐᵒᵖ N' where
+  smul_comm r' z n := show r' • z.unop • n = z.unop • r' • n from smul_comm _ _ _
+
+def _root_.LinearMap.toMulOpposite (l : M →ₗ[R'] N) : M →ₗ[R'ᵐᵒᵖ] N where
```

Comment:
This should be stated with `IsCentralScalar`, and not the instances above.

Reply:
This definition is no longer needed